File: drift/crypto/ratchet.py
"""
drift.crypto.ratchet — Signal Double Ratchet (Phase 2)

Reference: https://signal.org/docs/specifications/doubleratchet/

Provides per-message forward secrecy and post-compromise security on top of
the Phase 0/1 building blocks:

  - DH ratchet   — each party advertises a fresh X25519 ratchet public key in
                   every message header; whenever the peer's key changes, both
                   sides turn the DH ratchet, mixing a new shared secret into
                   the root key.
  - Symmetric ratchet — within a chain, each message key is derived from a
                   one-way KDF step on the chain key, then the chain key is
                   advanced and the old one discarded. Past message keys can
                   never be reconstructed from a later chain key.

KDFs
----
Every KDF here is HKDF-SHA256 from the `cryptography` library. We never write
our own HMAC or KDF. The root KDF follows the spec (salt = root key, IKM = DH
output); the chain KDF derives 64 bytes from the chain key and splits them into
(next chain key, message key) — equivalent in role to the spec's HMAC chain KDF
but expressed purely through HKDF.

All AEAD is the project's XChaCha20-Poly1305 `encrypt`/`decrypt`; the serialized
header is bound as associated data, so tampering with a header is rejected.
`InvalidTag` is always allowed to propagate (project iron rule).
"""

from __future__ import annotations

from dataclasses import dataclass, field

from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from drift.crypto import Keypair, decrypt, encrypt

# Maximum number of message keys we will skip (and cache) within a single chain
# before treating a header as hostile. Bounds memory against a flood of
# large-gap message numbers.
MAX_SKIP = 1000

_DH_PUB_LEN = 32
_COUNTER_LEN = 4
_HEADER_LEN = _DH_PUB_LEN + 2 * _COUNTER_LEN


class RatchetError(Exception):
    """Raised on a protocol violation (e.g. too many skipped messages)."""


# ---------------------------------------------------------------------------
# Message header
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Header:
    """
    Per-message ratchet header, sent in the clear alongside the ciphertext.

      dh — our current ratchet public key (32 bytes)
      pn — number of messages in the *previous* sending chain
      n  — this message's number in the current sending chain
    """
    dh: bytes
    pn: int
    n: int

    def to_bytes(self) -> bytes:
        """Canonical wire encoding: dh(32) || pn(4 BE) || n(4 BE)."""
        return (
            self.dh
            + self.pn.to_bytes(_COUNTER_LEN, "big")
            + self.n.to_bytes(_COUNTER_LEN, "big")
        )

    @classmethod
    def from_bytes(cls, raw: bytes) -> Header:
        if len(raw) != _HEADER_LEN:
            raise ValueError(f"header must be {_HEADER_LEN} bytes, got {len(raw)}")
        dh = raw[:_DH_PUB_LEN]
        pn = int.from_bytes(raw[_DH_PUB_LEN:_DH_PUB_LEN + _COUNTER_LEN], "big")
        n = int.from_bytes(raw[_DH_PUB_LEN + _COUNTER_LEN:], "big")
        return cls(dh=dh, pn=pn, n=n)


# ---------------------------------------------------------------------------
# Ratchet state
# ---------------------------------------------------------------------------

@dataclass
class RatchetState:
    """
    The full mutable state of one party's Double Ratchet session.

      root_key            — RK,  the root key (32 bytes)
      sending_chain_key   — CKs, current sending chain key (None until ready)
      receiving_chain_key — CKr, current receiving chain key (None until ready)
      message_keys        — MKSKIPPED: cache of skipped message keys, keyed by
                            (their_ratchet_pub, message_number)
      ratchet_keypair     — DHs, our current X25519 ratchet keypair
      their_ratchet_pub   — DHr, the peer's current ratchet public key
      send_count          — Ns, messages sent in the current sending chain
      recv_count          — Nr, messages received in the current receiving chain
      prev_send_count     — PN, length of the previous sending chain
    """
    root_key: bytes
    sending_chain_key: bytes | None
    receiving_chain_key: bytes | None
    ratchet_keypair: Keypair
    their_ratchet_pub: bytes | None
    send_count: int = 0
    recv_count: int = 0
    prev_send_count: int = 0
    message_keys: dict[tuple[bytes, int], bytes] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# KDFs (HKDF-SHA256 only)
# ---------------------------------------------------------------------------

def _kdf_rk(root_key: bytes, dh_out: bytes) -> tuple[bytes, bytes]:
    """
    Root KDF — per spec: HKDF with salt = root key, IKM = DH output.
    Returns (new_root_key, chain_key).
    """
    out = HKDF(
        algorithm=SHA256(), length=64, salt=root_key, info=b"drift-ratchet-v1-rk"
    ).derive(dh_out)
    return out[:32], out[32:]


def _kdf_ck(chain_key: bytes) -> tuple[bytes, bytes]:
    """
    Chain KDF — one HKDF step over the chain key, split into
    (next_chain_key, message_key). One-way: a later chain key reveals
    nothing about earlier chain keys or message keys.
    """
    out = HKDF(
        algorithm=SHA256(), length=64, salt=None, info=b"drift-ratchet-v1-ck"
    ).derive(chain_key)
    return out[:32], out[32:]


# ---------------------------------------------------------------------------
# Initialisation
# ---------------------------------------------------------------------------

def init_sender(shared_secret: bytes, their_ratchet_pub: bytes) -> RatchetState:
    """
    Initialise the initiating party (Signal's "Alice").

    The initiator generates a fresh ratchet keypair and immediately turns the
    root ratchet using DH(our new key, their initial ratchet public key),
    producing the first sending chain. The initiator can send right away.
    """
    dhs = Keypair.generate()
    root_key, sending_chain_key = _kdf_rk(shared_secret, dhs.ecdh(their_ratchet_pub))
    return RatchetState(
        root_key=root_key,
        sending_chain_key=sending_chain_key,
        receiving_chain_key=None,
        ratchet_keypair=dhs,
        their_ratchet_pub=their_ratchet_pub,
    )


def init_receiver(shared_secret: bytes, our_ratchet_keypair: Keypair) -> RatchetState:
    """
    Initialise the responding party (Signal's "Bob").

    The responder holds the ratchet keypair whose public half the initiator
    used to bootstrap. It has no sending chain yet — it must receive the
    initiator's first message (and turn the DH ratchet) before it can send.
    """
    return RatchetState(
        root_key=shared_secret,
        sending_chain_key=None,
        receiving_chain_key=None,
        ratchet_keypair=our_ratchet_keypair,
        their_ratchet_pub=None,
    )


# ---------------------------------------------------------------------------
# Encrypt / decrypt
# ---------------------------------------------------------------------------

def ratchet_encrypt(state: RatchetState, plaintext: bytes) -> tuple[Header, bytes]:
    """
    Advance the sending chain and encrypt one message.

    Returns (header, ciphertext). The header carries our ratchet public key
    and message number; it is also bound as AEAD associated data.
    """
    if state.sending_chain_key is None:
        raise RatchetError(
            "no sending chain yet — receive a message before sending"
        )
    state.sending_chain_key, message_key = _kdf_ck(state.sending_chain_key)
    header = Header(
        dh=state.ratchet_keypair.public_bytes(),
        pn=state.prev_send_count,
        n=state.send_count,
    )
    state.send_count += 1
    ciphertext = encrypt(message_key, plaintext, associated_data=header.to_bytes())
    return header, ciphertext


def ratchet_decrypt(state: RatchetState, header: Header, ciphertext: bytes) -> bytes:
    """
    Decrypt one message, advancing / turning ratchets as needed.

    Handles out-of-order delivery via the skipped-message-key cache and turns
    the DH ratchet whenever the header advertises a new peer ratchet key.
    A genuine decryption (authentication) failure raises ``InvalidTag``.
    """
    # 1. A message we already skipped and cached.
    plaintext = _try_skipped_keys(state, header, ciphertext)
    if plaintext is not None:
        return plaintext

    # 2. New peer ratchet key → skip the rest of the old chain, then DH ratchet.
    if header.dh != state.their_ratchet_pub:
        _skip_message_keys(state, header.pn)
        _dh_ratchet(state, header)

    # 3. Skip up to this message's number in the current receiving chain.
    _skip_message_keys(state, header.n)

    # 4. Derive this message's key and decrypt. In normal operation the DH
    #    ratchet above guarantees a receiving chain; its absence means the
    #    chain key was erased (e.g. after secure deletion) — unrecoverable.
    if state.receiving_chain_key is None:
        raise RatchetError("no receiving chain key — message is unrecoverable")
    state.receiving_chain_key, message_key = _kdf_ck(state.receiving_chain_key)
    state.recv_count += 1
    return decrypt(message_key, ciphertext, associated_data=header.to_bytes())


# ---------------------------------------------------------------------------
# Internal ratchet mechanics
# ---------------------------------------------------------------------------

def _try_skipped_keys(
    state: RatchetState, header: Header, ciphertext: bytes
) -> bytes | None:
    """Use (and consume) a cached message key, if one matches this header."""
    key = (header.dh, header.n)
    message_key = state.message_keys.pop(key, None)
    if message_key is None:
        return None
    # InvalidTag propagates — a cached key that fails to authenticate is tamper.
    return decrypt(message_key, ciphertext, associated_data=header.to_bytes())


def _skip_message_keys(state: RatchetState, until: int) -> None:
    """
    Derive and cache message keys for the current receiving chain up to (but
    not including) message number ``until`` — covering out-of-order arrivals.
    """
    if state.recv_count + MAX_SKIP < until:
        raise RatchetError("too many skipped messages")
    if state.receiving_chain_key is None:
        return
    while state.recv_count < until:
        state.receiving_chain_key, message_key = _kdf_ck(state.receiving_chain_key)
        assert state.their_ratchet_pub is not None
        state.message_keys[(state.their_ratchet_pub, state.recv_count)] = message_key
        state.recv_count += 1


def _dh_ratchet(state: RatchetState, header: Header) -> None:
    """
    Turn the DH ratchet on receipt of a new peer ratchet key: derive a fresh
    receiving chain, generate a new local ratchet keypair, and derive a fresh
    sending chain.
    """
    state.prev_send_count = state.send_count
    state.send_count = 0
    state.recv_count = 0
    state.their_ratchet_pub = header.dh

    state.root_key, state.receiving_chain_key = _kdf_rk(
        state.root_key, state.ratchet_keypair.ecdh(header.dh)
    )
    state.ratchet_keypair = Keypair.generate()
    state.root_key, state.sending_chain_key = _kdf_rk(
        state.root_key, state.ratchet_keypair.ecdh(header.dh)
    )

File: drift-core/crypto/src/ratchet.rs
//! The Signal Double Ratchet, matching `drift.crypto.ratchet`.
//!
//! Forward secrecy + post-compromise security over the AEAD/KDF building blocks.
//! The decrypt path is *transactional* (audit H1): it runs on a snapshot and
//! commits back to the live state only after the message authenticates, so a
//! forged header naming a new DH key can drive the trial DH-ratchet without ever
//! corrupting a live session. `Error::InvalidTag` always propagates.

use std::collections::HashMap;

use crate::aead::{decrypt, encrypt_with_nonce};
use crate::entropy::Entropy;
use crate::identity::Keypair;
use crate::kdf::{derive_message_key, kdf_ck, kdf_rk};
use crate::{Error, Result};

/// Cap on skipped-and-cached message keys within a chain (flood bound).
pub const MAX_SKIP: u32 = 1000;

/// Domain separation for the bootstrap forward-secrecy secret (audit H3).
const FS_BOOTSTRAP_INFO: &[u8] = b"drift-ratchet-v1-fs-bootstrap";

const DH_PUB_LEN: usize = 32;
const COUNTER_LEN: usize = 4;
const HEADER_LEN: usize = DH_PUB_LEN + 2 * COUNTER_LEN;

/// Per-message ratchet header: `dh(32) || pn(4 BE) || n(4 BE)`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Header {
    pub dh: [u8; 32],
    pub pn: u32,
    pub n: u32,
}

impl Header {
    pub fn to_bytes(&self) -> [u8; HEADER_LEN] {
        let mut out = [0u8; HEADER_LEN];
        out[..DH_PUB_LEN].copy_from_slice(&self.dh);
        out[DH_PUB_LEN..DH_PUB_LEN + COUNTER_LEN].copy_from_slice(&self.pn.to_be_bytes());
        out[DH_PUB_LEN + COUNTER_LEN..].copy_from_slice(&self.n.to_be_bytes());
        out
    }

    pub fn from_bytes(raw: &[u8]) -> Result<Header> {
        if raw.len() != HEADER_LEN {
            return Err(Error::Malformed("ratchet header wrong length"));
        }
        let mut dh = [0u8; 32];
        dh.copy_from_slice(&raw[..DH_PUB_LEN]);
        let pn = u32::from_be_bytes(
            raw[DH_PUB_LEN..DH_PUB_LEN + COUNTER_LEN]
                .try_into()
                .unwrap(),
        );
        let n = u32::from_be_bytes(raw[DH_PUB_LEN + COUNTER_LEN..].try_into().unwrap());
        Ok(Header { dh, pn, n })
    }
}

/// One party's full Double Ratchet state.
#[derive(Clone)]
pub struct RatchetState {
    pub root_key: [u8; 32],
    pub sending_chain_key: Option<[u8; 32]>,
    pub receiving_chain_key: Option<[u8; 32]>,
    pub ratchet_keypair: Keypair,
    pub their_ratchet_pub: Option<[u8; 32]>,
    pub send_count: u32,
    pub recv_count: u32,
    pub prev_send_count: u32,
    pub message_keys: HashMap<([u8; 32], u32), [u8; 32]>,
}

/// Initialise the initiator ("Alice"): generate a ratchet keypair and turn the
/// root ratchet against the peer's initial ratchet public key, ready to send.
pub fn init_sender(
    shared_secret: &[u8; 32],
    their_ratchet_pub: &[u8; 32],
    entropy: &mut impl Entropy,
) -> RatchetState {
    let dhs = entropy.ratchet_keypair();
    let dh_out = dhs.ecdh(their_ratchet_pub);
    let (root_key, sending_chain_key) = kdf_rk(shared_secret, &dh_out);
    RatchetState {
        root_key,
        sending_chain_key: Some(sending_chain_key),
        receiving_chain_key: None,
        ratchet_keypair: dhs,
        their_ratchet_pub: Some(*their_ratchet_pub),
        send_count: 0,
        recv_count: 0,
        prev_send_count: 0,
        message_keys: HashMap::new(),
    }
}

/// Initialise the responder ("Bob"): holds the ratchet keypair whose public half
/// the initiator bootstrapped against; no sending chain until it receives.
pub fn init_receiver(shared_secret: &[u8; 32], our_ratchet_keypair: Keypair) -> RatchetState {
    RatchetState {
        root_key: *shared_secret,
        sending_chain_key: None,
        receiving_chain_key: None,
        ratchet_keypair: our_ratchet_keypair,
        their_ratchet_pub: None,
        send_count: 0,
        recv_count: 0,
        prev_send_count: 0,
        message_keys: HashMap::new(),
    }
}

/// Advance the sending chain and encrypt one message.
pub fn ratchet_encrypt(
    state: &mut RatchetState,
    plaintext: &[u8],
    entropy: &mut impl Entropy,
) -> Result<(Header, Vec<u8>)> {
    let ck = state.sending_chain_key.ok_or(Error::Ratchet(
        "no sending chain yet — receive a message before sending",
    ))?;
    let (next_ck, message_key) = kdf_ck(&ck);
    state.sending_chain_key = Some(next_ck);
    let header = Header {
        dh: state.ratchet_keypair.public_bytes(),
        pn: state.prev_send_count,
        n: state.send_count,
    };
    state.send_count += 1;
    let nonce = entropy.aead_nonce();
    let ct = encrypt_with_nonce(&message_key, &nonce, plaintext, &header.to_bytes());
    Ok((header, ct))
}

/// Decrypt one message, turning ratchets as needed. Transactional: mutates a
/// clone and commits only on a successful authenticated decrypt (audit H1).
pub fn ratchet_decrypt(
    state: &mut RatchetState,
    header: &Header,
    ciphertext: &[u8],
    root_mix: Option<&[u8]>,
    entropy: &mut impl Entropy,
) -> Result<Vec<u8>> {
    let mut trial = state.clone();
    let plaintext = decrypt_into(&mut trial, header, ciphertext, root_mix, entropy)?;
    *state = trial;
    Ok(plaintext)
}

fn decrypt_into(
    state: &mut RatchetState,
    header: &Header,
    ciphertext: &[u8],
    root_mix: Option<&[u8]>,
    entropy: &mut impl Entropy,
) -> Result<Vec<u8>> {
    // 1. A message we already skipped and cached.
    if let Some(pt) = try_skipped_keys(state, header, ciphertext)? {
        return Ok(pt);
    }

    // 2. New peer ratchet key → skip the rest of the old chain, then DH ratchet.
    if Some(header.dh) != state.their_ratchet_pub {
        if let Some(mix) = root_mix {
            if state.their_ratchet_pub.is_none() {
                // Bootstrap-only forward-secrecy fold (audit H3).
                state.root_key = derive_message_key(mix, Some(&state.root_key), FS_BOOTSTRAP_INFO);
            }
        }
        skip_message_keys(state, header.pn)?;
        dh_ratchet(state, header, entropy);
    }

    // 3. Skip up to this message's number in the current receiving chain.
    skip_message_keys(state, header.n)?;

    // 4. Derive this message's key and decrypt.
    let rck = state.receiving_chain_key.ok_or(Error::Ratchet(
        "no receiving chain key — message is unrecoverable",
    ))?;
    let (next_rck, message_key) = kdf_ck(&rck);
    state.receiving_chain_key = Some(next_rck);
    state.recv_count += 1;
    decrypt(&message_key, ciphertext, &header.to_bytes())
}

fn try_skipped_keys(
    state: &mut RatchetState,
    header: &Header,
    ciphertext: &[u8],
) -> Result<Option<Vec<u8>>> {
    let key = (header.dh, header.n);
    match state.message_keys.remove(&key) {
        Some(mk) => Ok(Some(decrypt(&mk, ciphertext, &header.to_bytes())?)),
        None => Ok(None),
    }
}

fn skip_message_keys(state: &mut RatchetState, until: u32) -> Result<()> {
    if state.recv_count + MAX_SKIP < until {
        return Err(Error::Ratchet("too many skipped messages"));
    }
    let their = match state.their_ratchet_pub {
        Some(t) => t,
        None => return Ok(()),
    };
    let mut rck = match state.receiving_chain_key {
        Some(c) => c,
        None => return Ok(()),
    };
    while state.recv_count < until {
        let (next_rck, message_key) = kdf_ck(&rck);
        rck = next_rck;
        state
            .message_keys
            .insert((their, state.recv_count), message_key);
        state.recv_count += 1;
    }
    state.receiving_chain_key = Some(rck);
    Ok(())
}

fn dh_ratchet(state: &mut RatchetState, header: &Header, entropy: &mut impl Entropy) {
    state.prev_send_count = state.send_count;
    state.send_count = 0;
    state.recv_count = 0;
    state.their_ratchet_pub = Some(header.dh);

    let dh_out = state.ratchet_keypair.ecdh(&header.dh);
    let (root_key, receiving_chain_key) = kdf_rk(&state.root_key, &dh_out);
    state.root_key = root_key;
    state.receiving_chain_key = Some(receiving_chain_key);

    state.ratchet_keypair = entropy.ratchet_keypair();
    let dh_out2 = state.ratchet_keypair.ecdh(&header.dh);
    let (root_key2, sending_chain_key) = kdf_rk(&state.root_key, &dh_out2);
    state.root_key = root_key2;
    state.sending_chain_key = Some(sending_chain_key);
}
